### PR TITLE
Kubeadm - Added initial support for Windows worker nodes to join cluster using kubeadm

### LINKED ISSUE #364: Build kubeadm for Windows Server
## Is this a BUG REPORT or FEATURE REQUEST?

FEATURE REQUEST

## Versions

**kubeadm version** (use `kubeadm version`): `kubeadm version: &version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.2",
GitCommit:"922a86cfcd65915a9b2f69f3f193b8907d741d9c", GitTreeState:"clean",
BuildDate:"2017-07-21T08:08:00Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}`

**Environment**:
- **Kubernetes version** (use `kubectl version`): `Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.2", GitCommit:"922a86cfcd65915a9b2f69f3f193b8907d741d9c", GitTreeState:"clean", BuildDate:"2017-07-21T08:23:22Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}`
- **Cloud provider or hardware configuration**: Hyper-V and bare metal
- **OS** (e.g. from /etc/os-release): Ubuntu 16.04.2 LTS, Windows Server 2016
- **Kernel** (e.g. `uname -a`): `4.4.0-87-generic`
- **Others**: N/A

## What happened?

The documentation indicates that `kubeadm` is the recommended tool to use for creating a new Kubernetes cluster, and it seems to be working fine on Linux despite the beta warnings. However, there isn't currently a Windows build for the tool, so I can't connect Windows nodes to the cluster. It seems like this should be possible if I write a script to run the join steps of https://kubernetes.io/docs/admin/kubeadm/, but it would be a lot easier (and far more user-friendly) if I could use `kubeadm join` on Windows as well.

Ideally `kubeadm` will become entirely cross-platform in the future, but as Windows control planes aren't supported yet that will likely take a lot more work.

### PR DESCRIPTION
**Pull request by bsteciuk**
**What this PR does / why we need it**:
This PR adds initial support for adding a Windows worker node to a Kubernetes cluster with kubeadm.  Also adds Windows build of kubeadm to node build targets. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes https://github.com/kubernetes/kubeadm/issues/364 

**Special notes for your reviewer**:

Depends on #53730 

**Release note**:

```release-note
kubeadm: Add support for adding a Windows node
```

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go
Comment (luxas) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go]: Please don't put `if windows { } else { }` generally in the code here. Use generic path helper funcs where there is a difference between platforms

File: test/e2e_node/system/types.go
Comment (luxas) [inline test/e2e_node/system/types.go]: could we have a map or something that maps `GOOS` and `SysSpec` possibly?
That would make this more reusable I guess...

File: test/e2e_node/system/kernel_validator_helper.go
Comment (luxas) [inline test/e2e_node/system/kernel_validator_helper.go]: Godoc comments everywhere please

File: test/e2e_node/system/kernel_validator_helper.go
Comment (luxas) [inline test/e2e_node/system/kernel_validator_helper.go]: 2017 :)

File: test/e2e_node/system/kernel_validator.go
Comment (luxas) [inline test/e2e_node/system/kernel_validator.go]: nit: `// Foo bar baz`

File: test/e2e_node/system/docker_validator.go
Comment (luxas) [inline test/e2e_node/system/docker_validator.go]: Can this be done as a map to avoid the `if else` statement please?

File: pkg/util/initsystem/initsystem.go
Comment (luxas) [inline pkg/util/initsystem/initsystem.go]: Godoc comments

File: cmd/kubeadm/app/preflight/checks.go
Comment (luxas) [inline cmd/kubeadm/app/preflight/checks.go]: can as well do `if linux {`, that's more correct I guess

File: cmd/kubeadm/app/preflight/checks.go
Comment (luxas) [inline cmd/kubeadm/app/preflight/checks.go]: Any other method of checking admin privs that might be more obvious?

File: cmd/kubeadm/app/preflight/checks.go
Comment (luxas) [inline cmd/kubeadm/app/preflight/checks.go]: windowsPrivilegedCheck() and linuxPrivilegedCheck() possibly?

File: cmd/kubeadm/app/preflight/checks.go
Comment (luxas) [inline cmd/kubeadm/app/preflight/checks.go]: can we dedupe and make this code flow a bit cleaner?

**Review by luxas — COMMENTED** [review]
Thanks! Great progress, I'm so happy to see you working on this :smile: :+1: 

Left a couple of comments. Mostly around code ordering / etc.

I want you to split all `test/e2e_node` changes into a separate, generic PR for review by sig-node though, and then merge this after that one.

**Comment by bsteciuk** [comment]
Created #53730 for test/e2e_node changes as requested.

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go
Comment (kad) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go]: nit: if std "runtime" not used in this file, do we need now to rename that import ?

File: cmd/kubeadm/app/preflight/checks.go
Comment (kad) [inline cmd/kubeadm/app/preflight/checks.go]: What do you think about splitting this into more Go idiomatic way to `checks_windows.go` and have it build only in case of windows ?

File: cmd/kubeadm/app/util/path.go
Comment (kad) [inline cmd/kubeadm/app/util/path.go]: same thing: maybe split it to path_windows.go and build only on windows ?

Comment (bsteciuk) [reply cmd/kubeadm/app/util/path.go]: @kad would it make more sense just to split up defaults.go for necessary OS specific defaults (at the moment just one path), and get rid of the NormalizePath() func?

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go
Comment (bsteciuk) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go]: yup, runtime had been in there, but is no longer, updating

File: cmd/kubeadm/app/preflight/checks.go
Comment (bsteciuk) [reply cmd/kubeadm/app/preflight/checks.go]: :+1:

File: cmd/kubeadm/app/util/path.go
Comment (kad) [reply cmd/kubeadm/app/util/path.go]: I don't know. probably @luxas has opinion about that ?

File: pkg/util/initsystem/initsystem.go
Comment (brendandburns) [inline pkg/util/initsystem/initsystem.go]: Not going to make you do this, since this was already this way, but it would probably be good to switch to `util.exec` for testability...

File: test/e2e_node/system/kernel_validator.go
Comment (brendandburns) [inline test/e2e_node/system/kernel_validator.go]: probably worth doing the `*_windows.go` here too instead of a function pointer.

**Comment by brendandburns** [comment]
Awesome to see this start to come together. Some minor comments.

File: test/e2e_node/system/kernel_validator.go
Comment (bsteciuk) [reply test/e2e_node/system/kernel_validator.go]: @brendanburns yup, this was already done in the other PR #53730 made for the e2e_node portions.

**Comment by luxas** [comment]
/hold

^ means we're waiting for https://github.com/kubernetes/kubernetes/pull/53730 to get merged first

**Comment by brendandburns** [comment]
/lgtm
/approve

Approved #53730, I'll lift the hold once it merges.

File: cmd/kubeadm/app/preflight/checks.go
Comment (luxas) [inline cmd/kubeadm/app/preflight/checks.go]: I'm not sure why you made this extra variable, things seem to be fine without it (using `runtime.GOOS` directly in this func)

File: cmd/kubeadm/app/preflight/checks.go
Comment (luxas) [inline cmd/kubeadm/app/preflight/checks.go]: Can we switch this to `runtime.GOOS == "linux"` instead?

File: cmd/kubeadm/app/preflight/checks.go
Comment (luxas) [inline cmd/kubeadm/app/preflight/checks.go]: I would expect the system verification check to support windows in this validator; can you fix that?

File: cmd/kubeadm/app/preflight/checks_windows.go
Comment (luxas) [inline cmd/kubeadm/app/preflight/checks_windows.go]: follow the normal
```go
import (
    "fmt"
    "strings"
...
)
```
pattern please

File: pkg/util/initsystem/initsystem.go
Comment (luxas) [reply pkg/util/initsystem/initsystem.go]: maybe @bsteciuk can follow up later?

**Review by luxas — COMMENTED** [review]
Thanks @bsteciuk.
I left some nits only, soon this is ready to be merged!

File: cmd/kubeadm/app/preflight/checks.go
Comment (bsteciuk) [reply cmd/kubeadm/app/preflight/checks.go]: @luxas, I'm not sure about this one.  The current check runs `uname` and verifies that it returns `sysSpec.OS`.  For unix, it verifies it returns `Linux` as the unix variant, as opposed to SunOS, or FreeBSD, etc.  I'm not sure there's an equivalent command or even the same "variant" concept in Windows.

File: cmd/kubeadm/app/preflight/checks_windows.go
Comment (bsteciuk) [reply cmd/kubeadm/app/preflight/checks_windows.go]: fixed.

File: cmd/kubeadm/app/preflight/checks.go
Comment (bsteciuk) [reply cmd/kubeadm/app/preflight/checks.go]: fixed

File: cmd/kubeadm/app/preflight/checks.go
Comment (bsteciuk) [reply cmd/kubeadm/app/preflight/checks.go]: fixed

**Review by luxas — APPROVED** [review]
@bsteciuk Only a commit squash away from LGTM ^^

**Review by luxas — APPROVED** [review]
/lgtm
/retest

**Comment by luxas** [comment]
flakes on https://github.com/kubernetes/kubernetes/issues/54873

**Pull request merged**
